// homey-api/lib/HomeyAPI/HomeyAPIV3/RealtimeConsumer.js
'use strict';

class RealtimeConsumer {
  constructor({
    subscribe,
    getUri,
    debug,
    onConnect = () => {},
    onDisconnect = () => {},
    onReconnect = () => {},
    onReconnectError = () => {},
    onEvent = () => {},
    setConnected = () => {},
  }) {
    this.__subscribe = subscribe;
    this.__getUri = getUri;
    this.__debug = debug;
    this.__onConnect = onConnect;
    this.__onDisconnect = onDisconnect;
    this.__onReconnect = onReconnect;
    this.__onReconnectError = onReconnectError;
    this.__onEvent = onEvent;
    this.__setConnected = setConnected;

    this.__subscriptionPromise = null;
    this.__connectPromise = null;
    this.__disconnectPromise = null;
  }

  async connect() {
    this.__debug('connect');

    try {
      await Promise.resolve();
      await this.__disconnectPromise;
      // eslint-disable-next-line no-empty
    } catch (err) {}

    if (this.__connectPromise) {
      await this.__connectPromise;
      return;
    }

    if (this.__subscriptionPromise) {
      await this.__subscriptionPromise;
      return;
    }

    const connectPromise = Promise.resolve().then(async () => {
      if (!this.__subscriptionPromise) {
        this.__subscriptionPromise = Promise.resolve(
          this.__subscribe(this.__getUri(), {
            onConnect: () => {
              this.__debug('onConnect');
              this.__setConnected(true);
              this.__onConnect();
            },
            onDisconnect: (reason) => {
              this.__debug(`onDisconnect Reason:${reason}`);
              this.__setConnected(false);
              this.__onDisconnect(reason);
            },
            onReconnect: () => {
              this.__debug('onReconnect');
              this.__setConnected(true);
              this.__onReconnect();
            },
            onReconnectError: (err) => {
              this.__debug('onReconnectError', err.message);
              this.__onReconnectError(err);
            },
            onEvent: (event, data) => {
              this.__debug('onEvent', event, data);
              this.__onEvent(event, data);
            },
          })
        );
      }

      await this.__subscriptionPromise;
    });

    this.__connectPromise = connectPromise;

    connectPromise
      .catch(() => {
        this.__subscriptionPromise = null;
      })
      .finally(() => {
        if (this.__connectPromise === connectPromise) {
          this.__connectPromise = null;
        }
      });

    await connectPromise;
  }

  async disconnect() {
    this.__debug('disconnect');

    try {
      await this.__connectPromise;
      // eslint-disable-next-line no-empty
    } catch (err) {}

    if (this.__disconnectPromise) {
      await this.__disconnectPromise;
      return;
    }

    const disconnectPromise = Promise.resolve().then(async () => {
      this.__setConnected(false);

      // A failed connect clears __subscriptionPromise in connect(). In that case
      // disconnect() becomes a safe no-op because there is no subscription handle
      // left to unsubscribe.
      if (this.__subscriptionPromise) {
        await this.__subscriptionPromise
          .then((subscription) => subscription.unsubscribe())
          .catch((err) => this.__debug('Error Disconnecting:', err));
      }
    });

    this.__disconnectPromise = disconnectPromise;

    disconnectPromise
      .catch(() => {})
      .finally(() => {
        if (this.__disconnectPromise === disconnectPromise) {
          this.__disconnectPromise = null;
        }

        this.__subscriptionPromise = null;
      });

    await disconnectPromise;
  }
}

module.exports = RealtimeConsumer;
